> ### Note on Labs and Assignments:
>
> 🔧 Look for the **wrench emoji** 🔧 — it highlights where you're expected to take action!
>
> These sections are graded and are not optional.
>

# IS 4487 Lab 6: Data Cleaning

## Outline

- Load and inspect a new dataset (Megatelco)
- Fix column names and data types
- Handle missing values
- Remove duplicate rows
- Review and remove outliers
- Reflect on data quality

In this lab, we’ll clean the data to get it ready for transformations and analysis.

We will continue working with this dataset in **Lab 7**, where we will create new features and apply transformations.

<a href="https://colab.research.google.com/github/Stan-Pugsley/is_4487_base/blob/main/Labs/lab_06_data_cleaning.ipynb" target="_parent">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

## Megatelco Data Dictionary

 DEMOGRAPHIC VARIABLES:
 - College - has the customer attended some college (one, zero)
 - Income - annual income of customer
 - House - estimated price of the customer's home (if applicable)

 USAGE VARIABLES:
 - Data Overage Mb - Average number of megabytes that the customer used in excess of the plan limit (over last 12 months)
 - Data Leftover Mb - Average number of megabytes that the customer use was below the plan limit (over last 12 months)
 - Data Mb Used - Average number of megabytes used per month (over last 12 months)
 - Text Message Count - Average number of texts per month (over last 12 months)
 - Over 15 Minute Calls Per Month - Average number of calls over 15 minutes in duration per month (over last 12 months)
 - Average Call Duration- Average call duration (over last 12 months)

PHONE VARIABLES:
 - Operating System - Current operating system of phone
 - Handset Price - Retail price of the phone used by the customer

ATTITUDINAL VARIABLES:
 - Reported Satisfaction - Survey response to "How satisfied are you with your current phone plan?" (high, med, low)
 - Reported Usage Level - Survey response to "How much do you use your phone?" (high, med, low)
 - Considering Change of Plan - Survey response to "Are you currently planning to change companies when your contract expires?" (high, med, low)

OTHER VARIABLES
 - Leave - Did this customer churn with the last contract expiration? (LEAVE, STAY)
 - ID - Customer identifier

In [86]:
import pandas as pd

url = "https://raw.githubusercontent.com/Stan-Pugsley/is_4487_base/812e9f15c357a5657a2795631fcaa9d9363cb417/DataSets/megatelco_leave_survey_data_cleaning_v2.csv"
df = pd.read_csv(url)

df.head()

college    income  data_overage_mb  data_leftover_mb  data_mb_used  \
0     one  403137.0               70               0.0        6605.0   
1    zero  129700.0               67              16.0        6028.0   
2    zero   69741.0               60               0.0        1482.0   
3     one  377572.0                0              22.0        3005.0   
4    zero  382080.0                0               0.0        1794.0   

   text_message_count   house  handset_price  over_15mins_calls_per_month  \
0                 199  841317          653.0                          5.0   
1                 134  476664         1193.0                          5.0   
2                 176  810225         1037.0                          3.0   
3                 184  826967         1161.0                          0.0   
4                  74  951896         1023.0                          0.0   

   average_call_duration reported_satisfaction reported_usage_level  \
0                    8.0                   low                  low   
1                    5.0                   low                  low   
2                    8.0                   low                  low   
3                    5.0                   low                  low   
4                   14.0                   low                  low   

  considering_change_of_plan  leave     id operating_system  
0                        yes  LEAVE   8183          Android  
1                        yes  LEAVE  12501              IOS  
2                        yes   STAY   7425              IOS  
3                         no  LEAVE  13488              IOS  
4                        yes   STAY  11389              IOS

In [39]:
# create a copy of your dataset for use in part 4
copied_df = df.copy(deep=True)

## Part 1: Review Column Names and Structure

Think about:

- Are column names consistent (lowercase, no spaces)?
- Are there any typos or redundant labels?
- Do the rows and columns appear aligned? (Are all the columns the same size? Are all the rows the same size?)

Why this matters:
Inconsistent or messy column names can break code and make analysis harder to follow.




In [40]:
# Standardize column names: lowercase, no spaces
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

# Get column info and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15016 entries, 0 to 15015
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   college                      15016 non-null  object 
 1   income                       15006 non-null  float64
 2   data_overage_mb              15016 non-null  int64  
 3   data_leftover_mb             14916 non-null  float64
 4   data_mb_used                 14916 non-null  float64
 5   text_message_count           15016 non-null  int64  
 6   house                        15016 non-null  int64  
 7   handset_price                14916 non-null  float64
 8   over_15mins_calls_per_month  15013 non-null  float64
 9   average_call_duration        14916 non-null  float64
 10  reported_satisfaction        15016 non-null  object 
 11  reported_usage_level         15016 non-null  object 
 12  considering_change_of_plan   14201 non-null  object 
 13  leave           

In [41]:
# View descriptive statistics for numerical columns
df.describe()

income  data_overage_mb  data_leftover_mb  data_mb_used  \
count   15006.000000     15016.000000      14916.000000  14916.000000   
mean   242013.863455       153.430674         37.487664   4200.979686   
std    109627.859666       113.019892         28.052318   2203.802446   
min    -65000.000000         0.000000          0.000000    400.000000   
25%    147818.500000        54.000000         12.000000   2292.750000   
50%    241750.500000       151.000000         34.000000   4220.000000   
75%    336442.000000       242.000000         62.000000   6079.250000   
max    432000.000000       380.000000         89.000000   8000.000000   

       text_message_count         house  handset_price  \
count        15016.000000  1.501600e+04   14916.000000   
mean           135.946590  8.771293e+05     794.937249   
std             62.934783  2.870168e+05    1238.997927   
min             52.000000 -4.630000e+02    -200.000000   
25%             93.000000  6.444678e+05     498.000000   
50%            135.000000  8.762530e+05     777.000000   
75%            178.000000  1.098829e+06    1063.000000   
max           5000.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  15013.00000           14916.000000  15016.000000  
mean                      10.56551              10.060941  11856.541289  
std                        8.40421              41.188957   6812.183367  
min                        0.00000               1.000000      2.000000  
25%                        3.00000               5.000000   6135.000000  
50%                        9.00000              10.000000  11754.500000  
75%                       17.00000              14.000000  17390.500000  
max                       35.00000            5000.000000  25354.000000

### Inspect categorical variables
Note that `df.describe()` only provides summary for numeric and date type variables. For variables defined as object - which are string/text, some maybe categorical (with limited and fixed number of allowed values), and others may be true string (can be any text, not limited in value).

For variables defined as object that we suspect are categorical you will often want to know what values are included. We can do this using `df[colname].value_counts()`

In [42]:
display(df['college'].value_counts())
display(df['reported_satisfaction'].value_counts())
display(df['reported_usage_level'].value_counts())
display(df['considering_change_of_plan'].value_counts())
display(df['operating_system'].value_counts())
display(df['leave'].value_counts())

college
zero    7960
one     7056
Name: count, dtype: int64

reported_satisfaction
low     10850
high     3415
avg       751
Name: count, dtype: int64

reported_usage_level
low     12235
high     2536
avg       245
Name: count, dtype: int64

considering_change_of_plan
yes    9267
no     4934
Name: count, dtype: int64

operating_system
Android    7813
IOS        7203
Name: count, dtype: int64

leave
STAY     7532
LEAVE    7484
Name: count, dtype: int64

## Part 2: Convert Data Types

Before analysis, make sure each column is stored in the correct format. This helps avoid calculation errors, makes plotting smoother, and ensures models interpret the data correctly.

Think about:
- Are numbers accidentally stored as strings?
- Should repeated text values be converted to categories?
- Are "yes"/"no" columns better represented as binary (0/1) or categorical types?

Fixing data types now saves time and avoids issues later in your workflow.




In [43]:
# Check original data types
print("Original dtypes:\n", df.dtypes)

# Convert categorical text columns
df['college'] = df['college'].astype('category')
df['reported_satisfaction'] = df['reported_satisfaction'].astype('category')
df['operating_system'] = df['operating_system'].astype('category')

# Convert object/text columns with limited possible values with an order to ordinal categorical columns
df['reported_satisfaction'] = pd.Categorical(df['reported_satisfaction'], categories = ['low', 'avg', 'high'], ordered = True)

# Convert binary columns ('yes'/'no', 'LEAVE'/'STAY') to binary categorical
df['considering_change_of_plan'] = df['considering_change_of_plan'].astype('category')

# Check updated data types
print("\nUpdated dtypes:\n", df.dtypes)


Original dtypes:
 college                         object
income                         float64
data_overage_mb                  int64
data_leftover_mb               float64
data_mb_used                   float64
text_message_count               int64
house                            int64
handset_price                  float64
over_15mins_calls_per_month    float64
average_call_duration          float64
reported_satisfaction           object
reported_usage_level            object
considering_change_of_plan      object
leave                           object
id                               int64
operating_system                object
dtype: object

Updated dtypes:
 college                        category
income                          float64
data_overage_mb                   int64
data_leftover_mb                float64
data_mb_used                    float64
text_message_count                int64
house                             int64
handset_price                   float64
over_1

### 🔧 Try It Yourself – Part 2

1. Convert the `leave` column from "yes"/"no" to binary (`1`/`0`) and make it a **category**
2. Convert `reported_usage_level` to an **ordinal categorical** type
3. Convert `house` to an **integer** type
3. Use `.info()` to confirm the changes


In [44]:
# add code here 🔧
df['leave'] = df['leave'].map({'LEAVE': 1, 'STAY': 0}).astype('category')
df['reported_usage_level'] = pd.Categorical(df['reported_usage_level'], categories=['low', 'med', 'high'], ordered=True)
df['house'] = df['house'].astype('int')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15016 entries, 0 to 15015
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   college                      15016 non-null  category
 1   income                       15006 non-null  float64 
 2   data_overage_mb              15016 non-null  int64   
 3   data_leftover_mb             14916 non-null  float64 
 4   data_mb_used                 14916 non-null  float64 
 5   text_message_count           15016 non-null  int64   
 6   house                        15016 non-null  int64   
 7   handset_price                14916 non-null  float64 
 8   over_15mins_calls_per_month  15013 non-null  float64 
 9   average_call_duration        14916 non-null  float64 
 10  reported_satisfaction        15016 non-null  category
 11  reported_usage_level         14771 non-null  category
 12  considering_change_of_plan   14201 non-null  category
 13  l

## Part 3: Handle Missing Values

Missing data can break charts, skew stats, and disrupt models — so it needs to be handled carefully.

### Think about:
- Are the missing values random or patterned?
- Can we drop rows, or do we need to fill them?
- Should we use mean, median, or something else?

### Guidelines:
- Drop rows if there are only a few missing and the columns associated with them are essential to keep intact
- Use median to replace outliers in numeric columns
- Use 0 if the missing value means “none” (e.g. If the value was in response to: “do you have a history of chronic illness?” and the value was just left blank, we can assume that that blank just means “none” (the patient has no history of chronic illness))
- Use mode to replace categorical values

Cleaning missing values early avoids bigger problems later.

-----


**Note on `.loc` and Warnings** - When assigning values to a DataFrame, especially after filtering or copying, it's best to use `.loc` to avoid **`SettingWithCopyWarning`**. This ensures that you're updating the original data and not a temporary view of it.


In [45]:
# View missing value counts
print("Missing values per column:\n", df.isnull().sum())

# Fill 'handset_price' with median
df['handset_price'] = df['handset_price'].fillna(df['handset_price'].median())

# Drop rows with missing 'income' (if very few)
df = df.dropna(subset=['income']).copy()

# Fill missing 'data_leftover_mb' with 0 if it logically means no leftover data
df.loc[:, 'data_leftover_mb'] = df['data_leftover_mb'].fillna(0)

# Fill 'average_call_duration' with median if necessary
df.loc[:, 'average_call_duration'] = df['average_call_duration'].fillna(df['average_call_duration'].median())

# Fill 'data_mb_used' with median
df.loc[:, 'data_mb_used'] = df['data_mb_used'].fillna(df['data_mb_used'].median())

# Confirm updated missing values
print("\nMissing values after handling:\n", df.isnull().sum())


Missing values per column:
 college                          0
income                          10
data_overage_mb                  0
data_leftover_mb               100
data_mb_used                   100
text_message_count               0
house                            0
handset_price                  100
over_15mins_calls_per_month      3
average_call_duration          100
reported_satisfaction            0
reported_usage_level           245
considering_change_of_plan     815
leave                            0
id                               0
operating_system                 0
dtype: int64

Missing values after handling:
 college                          0
income                           0
data_overage_mb                  0
data_leftover_mb                 0
data_mb_used                     0
text_message_count               0
house                            0
handset_price                    0
over_15mins_calls_per_month      3
average_call_duration            0
reported_satisfa

### 🔧 Try It Yourself – Part 3


There are still some missing values in:

- `over_15mins_calls_per_month`
- `considering_change_of_plan`

Decide how to handle them based on what makes the most sense:

- Should you fill them with 0, the median, or something else?
- For categories, would a placeholder like "unknown" or the most common value work?
- Or is it better to drop those rows?

1. Write and execute code to handle the missing values in the remaining two columns.
2. Use `df.isnull().sum()` to confirm all missing values are handled.



In [46]:
# Add code here 🔧
df.loc[:, 'over_15mins_calls_per_month'] = df['over_15mins_calls_per_month'].fillna(0)
df.loc[:, 'considering_change_of_plan'] = df['considering_change_of_plan'].fillna('no')
print("\nMissing values after handling specific columns:\n", df.isnull().sum())


Missing values after handling specific columns:
 college                          0
income                           0
data_overage_mb                  0
data_leftover_mb                 0
data_mb_used                     0
text_message_count               0
house                            0
handset_price                    0
over_15mins_calls_per_month      0
average_call_duration            0
reported_satisfaction            0
reported_usage_level           245
considering_change_of_plan       0
leave                            0
id                               0
operating_system                 0
dtype: int64


## Part 4: Remove Duplicate Rows

Sometimes the same row appears more than once due to data entry or processing mistakes. It's important to check for and remove these duplicates.

Think about:
- Are there rows that are exactly the same?
- If duplicates exist, should you keep the first one, the last one, or none?

Why this matters:
Duplicate rows can inflate totals, distort statistics, and lead to inaccurate conclusions.


In [ ]:
# Check for exact duplicates
print(f"Number of duplicate rows: {df.duplicated().sum()}")

# Remove them, keeping the first occurrence
df = df.drop_duplicates()

# Confirm result
print(f"Remaining rows after removing duplicates: {len(df)}")

### 🔧 Try It Yourself – Part 4

1. Use `copied_df.duplicated().sum()` to count how many duplicates are in your dataset.
2. Try using `copied_df.drop_duplicates(keep='last')` instead — what changes?

### In Your Response:
1. Explore whether duplicate rows share the same ID or just values across all columns and comment on your observation.


In [60]:
# 🔧 Add code here:
copied_df.duplicated().sum()

np.int64(17)

In [63]:
df[df.duplicated(keep=False)]

college    income  data_overage_mb  data_leftover_mb  data_mb_used  \
27     one  324903.0              129               5.0        4056.0   
28     one  324903.0              129               5.0        4056.0   
30     one  103540.0                0               9.0        3072.0   
31     one  103540.0                0               9.0        3072.0   
32     one  103540.0                0               9.0        3072.0   
45    zero  290449.0                0              84.0        1184.0   
46     one  284098.0               65              20.0        4453.0   
47     one  384785.0              141              17.0        4644.0   
48    zero  246436.0               85               0.0        2665.0   
49    zero  400865.0              159               0.0        5656.0   
50    zero  331682.0               47              85.0        5754.0   
51    zero  346119.0              264               0.0        2429.0   
52    zero  290449.0                0              84.0        1184.0   
53     one  284098.0               65              20.0        4453.0   
54     one  384785.0              141              17.0        4644.0   
55    zero  246436.0               85               0.0        2665.0   
56    zero  400865.0              159               0.0        5656.0   
57    zero  331682.0               47              85.0        5754.0   
58    zero  346119.0              264               0.0        2429.0   
63    zero  195320.0                0              75.0        3791.0   
64    zero  195320.0                0              75.0        3791.0   
65     one  325119.0               66               0.0        6958.0   
66     one  414896.0              219              84.0        5008.0   
67     one  291877.0               56              13.0        6319.0   
68    zero  274179.0                0              13.0        7624.0   
69    zero  -65000.0                0               0.0        6444.0   
70     one  332075.0                0               0.0        4881.0   
71     one  325119.0               66               0.0        6958.0   
72     one  414896.0              219              84.0        5008.0   
73     one  291877.0               56              13.0        6319.0   
74    zero  274179.0                0              13.0        7624.0   
75    zero  -65000.0                0               0.0        6444.0   
76     one  332075.0                0               0.0        4881.0   

    text_message_count    house  handset_price  over_15mins_calls_per_month  \
27                 204   368142          705.0                         21.0   
28                 204   368142          705.0                         21.0   
30                 153   708011          704.0                          1.0   
31                 153   708011          704.0                          1.0   
32                 153   708011          704.0                          1.0   
45                  80  1222190          298.0                          1.0   
46                 107   646006          618.0                          1.0   
47                 202   624462          557.0                         25.0   
48                  94  1095959         1023.0                          3.0   
49                 130   771829         1292.0                         22.0   
50                  97  1121841          413.0                          3.0   
51                 201   829460          452.0                         24.0   
52                  80  1222190          298.0                          1.0   
53                 107   646006          618.0                          1.0   
54                 202   624462          557.0                         25.0   
55                  94  1095959         1023.0                          3.0   
56                 130   771829         1292.0                         22.0   
57                  97  1121841          413.0                          3.0   
58                 

In [64]:
df[df.duplicated(keep=False)].sort_values(by='id', ascending=False)

college    income  data_overage_mb  data_leftover_mb  data_mb_used  \
46     one  284098.0               65              20.0        4453.0   
53     one  284098.0               65              20.0        4453.0   
54     one  384785.0              141              17.0        4644.0   
47     one  384785.0              141              17.0        4644.0   
45    zero  290449.0                0              84.0        1184.0   
52    zero  290449.0                0              84.0        1184.0   
27     one  324903.0              129               5.0        4056.0   
28     one  324903.0              129               5.0        4056.0   
65     one  325119.0               66               0.0        6958.0   
71     one  325119.0               66               0.0        6958.0   
48    zero  246436.0               85               0.0        2665.0   
55    zero  246436.0               85               0.0        2665.0   
74    zero  274179.0                0              13.0        7624.0   
68    zero  274179.0                0              13.0        7624.0   
51    zero  346119.0              264               0.0        2429.0   
58    zero  346119.0              264               0.0        2429.0   
76     one  332075.0                0               0.0        4881.0   
70     one  332075.0                0               0.0        4881.0   
50    zero  331682.0               47              85.0        5754.0   
57    zero  331682.0               47              85.0        5754.0   
32     one  103540.0                0               9.0        3072.0   
30     one  103540.0                0               9.0        3072.0   
31     one  103540.0                0               9.0        3072.0   
73     one  291877.0               56              13.0        6319.0   
67     one  291877.0               56              13.0        6319.0   
63    zero  195320.0                0              75.0        3791.0   
64    zero  195320.0                0              75.0        3791.0   
66     one  414896.0              219              84.0        5008.0   
72     one  414896.0              219              84.0        5008.0   
56    zero  400865.0              159               0.0        5656.0   
49    zero  400865.0              159               0.0        5656.0   
69    zero  -65000.0                0               0.0        6444.0   
75    zero  -65000.0                0               0.0        6444.0   

    text_message_count    house  handset_price  over_15mins_calls_per_month  \
46                 107   646006          618.0                          1.0   
53                 107   646006          618.0                          1.0   
54                 202   624462          557.0                         25.0   
47                 202   624462          557.0                         25.0   
45                  80  1222190          298.0                          1.0   
52                  80  1222190          298.0                          1.0   
27                 204   368142          705.0                         21.0   
28                 204   368142          705.0                         21.0   
65                 158   570769         1303.0                          5.0   
71                 158   570769         1303.0                          5.0   
48                  94  1095959         1023.0                          3.0   
55                  94  1095959         1023.0                          3.0   
74                 214  1156854          325.0                          0.0   
68                 214  1156854          325.0                          0.0   
51                 201   829460          452.0                         24.0   
58                 201   829460          452.0                         24.0   
76                  55  1029606          704.0                          1.0   
70                  55  1029606          704.0                          1.0   
50                 

In [66]:
copied_df.drop_duplicates(keep='last').sort_values(by='id', ascending=True)

college    income  data_overage_mb  data_leftover_mb  data_mb_used  \
2657     zero  285790.0              191              22.0        4857.0   
1294      one  252093.0               46              39.0        6435.0   
3638      one  408112.0              209              33.0        7361.0   
4390     zero   71806.0                0               9.0        1056.0   
1240     zero   84910.0               63              88.0        6985.0   
...       ...       ...              ...               ...           ...   
6955      one  225867.0               72              40.0         803.0   
14814     one  177647.0              139              74.0        1818.0   
7847     zero  107055.0               69              70.0        5328.0   
9360     zero  195892.0               72              20.0        1454.0   
9155      one  385954.0              144              64.0        6015.0   

       text_message_count    house  handset_price  \
2657                  161   785947          266.0   
1294                  220  1093225          560.0   
3638                  168   906114          735.0   
4390                   80   612324          307.0   
1240                  213   863904          238.0   
...                   ...      ...            ...   
6955                  200  1220727         1299.0   
14814                  71  1182765          310.0   
7847                  212   901348         1033.0   
9360                  214  1062735          794.0   
9155                   86  1063737          360.0   

       over_15mins_calls_per_month  average_call_duration  \
2657                          19.0                    6.0   
1294                          14.0                    1.0   
3638                          11.0                    1.0   
4390                           3.0                    1.0   
1240                           4.0                    2.0   
...                            ...                    ...   
6955                           6.0                   16.0   
14814                          2.0                   10.0   
7847                           6.0                   11.0   
9360                           1.0                    8.0   
9155                           0.0                   13.0   

      reported_satisfaction reported_usage_level considering_change_of_plan  \
2657                    low                  low                        yes   
1294                    low                 high                        yes   
3638                    low                  low                        yes   
4390                   high                  low                         no   
1240                    low                 high                        yes   
...                     ...                  ...                        ...   
6955                    low                  low                        yes   
14814                   low                  low                        yes   
7847                    low                  low                         no   
9360                    low                  low                        yes   
9155                    low                  low                        yes   

       leave     id operating_system  
2657   LEAVE      2          Android  
1294   LEAVE      4          Android  
3638   LEAVE      5          Android  
4390    STAY      6          Android  
1240   LEAVE      8          Android  
...      ...    ...              ...  
6955    STAY  25349              IOS  
14814   STAY  25352          Android  
7847   LEAVE  25353              IOS  
9360    STAY  25354          Android  
9155    STAY  25354          Android  

[14999 rows x 16 columns]

### ✍️ Your Response: 🔧
1. I displayed all duplicate rows arranged by customer ID to better contrast them to each other. Looking at this data this way we can see that rows with identical customer ID also share indentical values in all other variables, indicating these rows are identical duplicates and that they were inputed into the data by error.

## Part 5: Identify and Remove Obvious Outliers

Outliers are values that fall far outside the normal range. They can come from data entry mistakes or rare cases.

- Use summary statistics or visual tools (like boxplots) to find them.
- Look for clearly unrealistic values — e.g., negative prices or extremely high data usage.
- Decide how to handle them:
  - Remove if they’re errors.
  - Keep if they’re valid but rare — or cap them if needed.

Outliers can distort averages, stretch visualizations, and mislead models, so it’s important to address them carefully.



In [67]:
# Remove negative or nonsensical values using business rules

# Example: remove rows where 'handset_price' is negative
df = df[df['handset_price'] >= 0]

# Example: remove rows with unusually long call durations
df = df[df['average_call_duration'] < 1000]

# Example: remove rows with extremely high text message counts
df = df[df['text_message_count'] < 1000]

# View shape after outlier filtering
print("Shape after removing obvious outliers:", df.shape)


Shape after removing obvious outliers: (15003, 16)


### 🔧 Try It Yourself – Part 5

1. Use `df.describe()` to look for columns with extreme minimum or maximum values.
2. Set a threshold for what you think is "too high" or "too low" for:
  - `data_mb_used`
  - `over_15mins_calls_per_month`
  - `income`
3. Remove those outliers using boolean filtering like `df = df[df['column'] < threshold]`

In [87]:
# 🔧 add code here:
df.describe()

income  data_overage_mb  data_leftover_mb  data_mb_used  \
count   15006.000000     15016.000000      14916.000000  14916.000000   
mean   242013.863455       153.430674         37.487664   4200.979686   
std    109627.859666       113.019892         28.052318   2203.802446   
min    -65000.000000         0.000000          0.000000    400.000000   
25%    147818.500000        54.000000         12.000000   2292.750000   
50%    241750.500000       151.000000         34.000000   4220.000000   
75%    336442.000000       242.000000         62.000000   6079.250000   
max    432000.000000       380.000000         89.000000   8000.000000   

       text_message_count         house  handset_price  \
count        15016.000000  1.501600e+04   14916.000000   
mean           135.946590  8.771293e+05     794.937249   
std             62.934783  2.870168e+05    1238.997927   
min             52.000000 -4.630000e+02    -200.000000   
25%             93.000000  6.444678e+05     498.000000   
50%            135.000000  8.762530e+05     777.000000   
75%            178.000000  1.098829e+06    1063.000000   
max           5000.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  15013.00000           14916.000000  15016.000000  
mean                      10.56551              10.060941  11856.541289  
std                        8.40421              41.188957   6812.183367  
min                        0.00000               1.000000      2.000000  
25%                        3.00000               5.000000   6135.000000  
50%                        9.00000              10.000000  11754.500000  
75%                       17.00000              14.000000  17390.500000  
max                       35.00000            5000.000000  25354.000000

In [90]:
df = df[df['income'] < 300000]
df = df[df['income'] > 0]
df = df[df['data_mb_used'] > 1000]
df = df[df['over_15mins_calls_per_month'] < 30]
df.describe()

income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    8740.000000      8740.000000       8682.000000   8740.000000   
mean   176131.271625       152.324828         37.095485   4500.520366   
std     71411.671476       112.802079         27.980208   2025.416706   
min     52491.000000         0.000000          0.000000   1001.000000   
25%    114042.500000        54.000000         12.000000   2726.750000   
50%    176478.500000       149.000000         34.000000   4509.500000   
75%    237290.750000       240.000000         62.000000   6243.250000   
max    299998.000000       380.000000         89.000000   8000.000000   

       text_message_count         house  handset_price  \
count         8740.000000  8.740000e+03    8677.000000   
mean           135.877803  8.764795e+05     803.439553   
std             71.148820  2.870511e+05    1587.550165   
min             52.000000 -4.630000e+02     215.000000   
25%             93.000000  6.456370e+05     496.000000   
50%            135.000000  8.736845e+05     777.000000   
75%            177.250000  1.099961e+06    1066.000000   
max           5000.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  8740.000000            8686.000000   8740.000000  
mean                     10.128604               9.661294  11858.170366  
std                       7.886608               5.227067   6826.209093  
min                       0.000000               1.000000      2.000000  
25%                       3.000000               5.000000   6062.750000  
50%                       9.000000              10.000000  11861.000000  
75%                      17.000000              14.000000  17412.750000  
max                      29.000000              19.000000  25354.000000

## Part 6: Handle Outliers Using Quantiles

Instead of removing outliers, we can limit their impact by capping (removing) extreme values — a method known as **Winsorizing**.

### How to Do It:
- Use `.quantile()` to identify the 1st and 99th percentiles (or other thresholds).
- Use `.clip()` to cap values within that range.

This keeps your dataset intact while reducing the influence of extreme values on your analysis or model.



In [91]:
# Calculate 1st and 99th percentiles for income
income_min, income_max = df['income'].quantile([0.01, 0.99])

# Use .loc to avoid SettingWithCopyWarning and ensure assignment modifies the original DataFrame
df.loc[:, 'income'] = df['income'].clip(lower=income_min, upper=income_max)

# Clip 'data_mb_used' to within 1st and 99th percentiles
usage_min, usage_max = df['data_mb_used'].quantile([0.01, 0.99])
df.loc[:, 'data_mb_used'] = df['data_mb_used'].clip(lower=usage_min, upper=usage_max)

# Clip 'average_call_duration' to reduce the effect of extreme outliers
call_min, call_max = df['average_call_duration'].quantile([0.01, 0.99])
df.loc[:, 'average_call_duration'] = df['average_call_duration'].clip(lower=call_min, upper=call_max)



### 🔧 Try It Yourself – Part 6

1. Use `.quantile([0.01, 0.99])` to find the range for:
  - `text_message_count`
  - `over_15mins_calls_per_month`
2. Apply `.clip(lower=..., upper=...)` to reduce the impact of those outliers

### In Your Response:
1. Compare the `.describe()` output before and after clipping and comment on what you observe


In [92]:
# 🔧 Add code here
df.describe()

income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    8740.000000      8740.000000       8682.000000   8740.000000   
mean   176133.237840       152.324828         37.095485   4500.466325   
std     71372.493994       112.802079         27.980208   2024.327207   
min     54723.750000         0.000000          0.000000   1059.000000   
25%    114042.500000        54.000000         12.000000   2726.750000   
50%    176478.500000       149.000000         34.000000   4509.500000   
75%    237290.750000       240.000000         62.000000   6243.250000   
max    297638.690000       380.000000         89.000000   7934.610000   

       text_message_count         house  handset_price  \
count         8740.000000  8.740000e+03    8677.000000   
mean           135.877803  8.764795e+05     803.439553   
std             71.148820  2.870511e+05    1587.550165   
min             52.000000 -4.630000e+02     215.000000   
25%             93.000000  6.456370e+05     496.000000   
50%            135.000000  8.736845e+05     777.000000   
75%            177.250000  1.099961e+06    1066.000000   
max           5000.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  8740.000000            8686.000000   8740.000000  
mean                     10.128604               9.661294  11858.170366  
std                       7.886608               5.227067   6826.209093  
min                       0.000000               1.000000      2.000000  
25%                       3.000000               5.000000   6062.750000  
50%                       9.000000              10.000000  11861.000000  
75%                      17.000000              14.000000  17412.750000  
max                      29.000000              19.000000  25354.000000

In [94]:
text_min, text_max = df['text_message_count'].quantile([0.01, 0.99])
df.loc[:, 'text_message_count'] = df['text_message_count'].clip(lower=text_min, upper=text_max)
calls_min, calls_max = df['over_15mins_calls_per_month'].quantile([0.01, 0.99])
df.loc[:, 'over_15mins_calls_per_month'] = df['over_15mins_calls_per_month'].clip(lower=calls_min, upper=calls_max)
df.describe()

income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    8740.000000      8740.000000       8682.000000   8740.000000   
mean   176133.237840       152.324828         37.095485   4500.466325   
std     71372.493994       112.802079         27.980208   2024.327207   
min     54723.750000         0.000000          0.000000   1059.000000   
25%    114042.500000        54.000000         12.000000   2726.750000   
50%    176478.500000       149.000000         34.000000   4509.500000   
75%    237290.750000       240.000000         62.000000   6243.250000   
max    297638.690000       380.000000         89.000000   7934.610000   

       text_message_count         house  handset_price  \
count         8740.000000  8.740000e+03    8677.000000   
mean           135.330892  8.764795e+05     803.439553   
std             48.507385  2.870511e+05    1587.550165   
min             53.000000 -4.630000e+02     215.000000   
25%             93.000000  6.456370e+05     496.000000   
50%            135.000000  8.736845e+05     777.000000   
75%            177.250000  1.099961e+06    1066.000000   
max            219.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  8740.000000            8686.000000   8740.000000  
mean                     10.120938               9.661294  11858.170366  
std                       7.868725               5.227067   6826.209093  
min                       0.000000               1.000000      2.000000  
25%                       3.000000               5.000000   6062.750000  
50%                       9.000000              10.000000  11861.000000  
75%                      17.000000              14.000000  17412.750000  
max                      28.000000              19.000000  25354.000000

### ✍️ Your Response: 🔧
1. Since text messages count was a prevously un altered varible I wanted to focus my comparion on that one. Taking the middle 98% of the data using the function above showed significant changes in the outcome, implying one or a small few of significant outliers. For example the original output has a standard deviation for text message count of 71, in the following output that excludes the top and bottom 1% the standard deviation goes down to 41. Another indication of an extreme outlier in this variable is the max value, the original value was 50000 far above the median or even the 75% percentile. in the new output the max value dropped to just 219, far close to the median.

## 🔧 Part 7: Reflection (100 words or less per question)

1. Which step fixed the most issues in the dataset?
2. What surprised you about the structure or values?
3. Do you feel this data is now ready for transformation in Lab 7?


### ✍️ Your Response: 🔧
1. Yes I think the [.01,.99] function in step 6 had the most effect, it acted as a easy one step solution to find and exclude extreme outliers without having to do more in depth analysis on the variables.
2. One thing that suprised me was the number of extreme outliers, text message count and over 15 mins call per month specifically had a few. The min value for income was negative too which suggested error in the data collection and it was interesting using the describe function to see all these anomolies.
3. Yes I beleive removing the duplicates and extreme outliers has made the data much more reliable now.

## Export Your Notebook to Submit in Canvas
- Use the instructions from Lab 1

In [95]:
!jupyter nbconvert --to html "lab_06_FugalEric.ipynb"

[NbConvertApp] Converting notebook lab_06_FugalEric.ipynb to html
[NbConvertApp] Writing 411898 bytes to lab_06_FugalEric.html
